### DISCUSSION TITLE
Above 12.6" Screenpad Plus Display (IPS, 1920x515 @ 60 Hz)

### DISCUSSION THREAD

**Discussion by MizhiDR** [Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh#58] [Post-Install Help]
Hello,12.6-inch Screenpad Plus display (IPS, 1920x515 @ 60 Hz) Can be displayed on macos14

**Comment by Maleficent-Magik** [comment]
Its a question or an affirmation ? 👀
Hello,12.6-inch Screenpad Plus display (IPS, 1920x515 @ 60 Hz) Can be displayed on macos14

**Reply by MizhiDR** [reply]
Sorry, I didn't express it correctly. This is a question.

**Comment by Maleficent-Magik** [comment]
The short answer is: For the moment, no, it's not working very, very well.

The more complex answer is: The project is slightly on stand-by, but you can take a look at what's been tried here:  <https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/4> https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/4

**Reply by MizhiDR** [reply]
Thank you for your efforts. Is the touchpad working properly?

**Reply by MizhiDR** [reply]
> The short answer is: For the moment, no, it's not working very, very well. The more complex answer is: The project is slightly on stand-by, but you can take a look at what's been tried here: <#4> #4
I tried the trackpad and it didn't work on macos14

**Reply by MizhiDR** [reply]
> > 简短的回答是：目前，不行，它工作得不是很好。更复杂的答案是：该项目略微处于待命状态，但您可以在此处查看已尝试的内容：< #4 > #4
> > 我尝试了触控板，但它在 macos14 上不起作用
I have added msgbuf=1048576

**Reply by MizhiDR** [reply]
https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/discussions/54#discussioncomment-10637161
Changing the values ​​and adding parameters as you said still doesn't work.

**Reply by Qonfused** [reply]
There doesn't exist a reliable fix for the screenpad at the moment, though there is a (very) hacky workaround in https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/4#issuecomment-1987709248 if you want to try your hand at it. The trackpad should work fine though how well touch-screen works may be hit or miss; you may want to try your luck reverting https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/commit/6fb418f160ccc1b27533e9fb651305e9cf5464b0 to see if GPIO mode works.

The main blocker for the screenpad is that any Kext-based patch can't be injected through OpenCore as it doesn't support SysKC injection (injecting the main drivers in `/System/Library/Extensions`); we need this for linking the IOGraphicsFamily or Coffee-Lake graphics drivers. It's the same blocker as in https://github.com/Qonfused/OSX-Hyper-V/issues/6, but if I can figure out a small patch to bypass the faulty initialization behavior I may be able to apply it as a kernel patch.

**Reply by MizhiDR** [reply]
I'm not talking about the screenpad at the moment，I want the trackpad to work
![60666316EEC6B41AE30894F30FCE155E](https://github.com/user-attachments/assets/308f7ba5-832f-4497-8cbb-2c28f9339f18)

**Reply by MizhiDR** [reply]
My ux481fl does not support screen touch

**Reply by Qonfused** [reply]
Ah, I had missed the post linking #54. It's very likely the GPIO pin used is locked in firmware and won't work with the current GPIO patch.

There is a workaround using the `-vi2c-force-polling` boot arg, but you'll need to edit the SSDT to disable the trackpad's GPIO pinning. I'd comment out the original return statement in lines 50-54 of `SSDT-GPIO.dsl` and instead write `Return (XCRS ())`:

```diff
# src/ACPI/SSDT-GPIO.dsl
...
  28 |    // GPIO - Trackpad
  29 |    Scope (_SB.PCI0.I2C1.ETPD)
  30 |    {
  31 |        Name (SBFG, ResourceTemplate ()
  32 |        {
  33 |            GpioInt (Level, ActiveLow, ExclusiveAndWake, PullDefault, 0x0000,
  34 |                "\\_SB.PCI0.GPI0", 0x00, ResourceConsumer, ,
  35 |                )
  36 |                {   // Pin list
  37 |                    0x0055
  38 |                }
  39 |        })
  40 |        
  41 |        Method (_CRS, 0, Serialized)
  42 |        {
  43 |            Name (SBFB, ResourceTemplate ()
  44 |            {
  45 |                I2cSerialBusV2 (0x0015, ControllerInitiated, 0x00061A80,
  46 |                    AddressingMode7Bit, "\\_SB.PCI0.I2C1",
  47 |                    0x00, ResourceConsumer, , Exclusive,
  48 |                    )
  49 |            })
- 50 |            If (_OSI ("Darwin"))
- 51 |            {
- 52 |                Return (ConcatenateResTemplate (SBFB, SBFG))
- 53 |            }
- 54 |            Else { Return (XCRS ()) }
+ 50 |            Return (XCRS ())
  51 |        }
  52 |    }
```

**Reply by MizhiDR** [reply]
Above Scope (_SB.PCI0.I2C1.ETPD)，Do you want to change it to your own touchpad address?

**Reply by Qonfused** [reply]
You would if it differs in the DSDT but it should match for all UX481FL variants since they use the same firmware.

**Reply by MizhiDR** [reply]
> 如果 DSDT 有所不同，您会这样做，但它应该与所有 UX481FL 变体匹配，因为它们使用相同的固件。


I tried your method but it still doesn’t work. Is it related to VoodooI2C.kext?

**Reply by MizhiDR** [reply]
> You would if it differs in the DSDT but it should match for all UX481FL variants since they use the same firmware.


I updated VoodooI2C.kext and used your Voodool2CHID

**Reply by Qonfused** [reply]
> > 如果 DSDT 有所不同，您会这样做，但它应该与所有 UX481FL 变体匹配，因为它们使用相同的固件。
> 
> I tried your method but it still doesn’t work. Is it related to VoodooI2C.kext?

It's possible, though VoodooI2C + VoodooI2CHID should match onto the device correctly with [VoodooInjector](https://github.com/Qonfused/VoodooInjector) (already included in the build). Within macOS, try running the below command and look for any `ETPD` devices in the output:
```bash
ioreg -rxn IOHIDInterface -k "VoodooI2CServices Supported" -t | grep -E 'IOACPIPlatformDevice|IOPCIDevice|VoodooI2CDeviceNub|VendorID|ProductID|HIDEventDriver  <class VoodooI2C|Interface  <class VoodooI2C' | cut -d "<" -f1
```

If there aren't any `ETPD` devices in the output, then the device itself hasn't been enabled in your DSDT. In either case though, I'd [dump your DSDT](https://dortania.github.io/Getting-Started-With-ACPI/Manual/dump.html#from-windows) on Windows so we can verify that it can correctly enable with the current SSDT-GPIO patch.

> > You would if it differs in the DSDT but it should match for all UX481FL variants since they use the same firmware.
> 
> I updated VoodooI2C.kext and used your VoodooI2CHID

You can also use the newer build of VoodooI2CHID from v2.9.1; my build back-ported changes to v2.8.

**Reply by MizhiDR** [reply]
> > > 如果 DSDT 有所不同，您会这样做，但它应该与所有 UX481FL 变体匹配，因为它们使用相同的固件。
> > 
> > 
> > I tried your method but it still doesn’t work. Is it related to VoodooI2C.kext?
> 
> It's possible, though VoodooI2C + VoodooI2CHID should match onto the device correctly with [VoodooInjector](https://github.com/Qonfused/VoodooInjector) (already included in the build). Within macOS, try running the below command and look for any `ETPD` devices in the output:
> 
> ```shell
> ioreg -rxn IOHIDInterface -k "VoodooI2CServices Supported" -t | grep -E 'IOACPIPlatformDevice|IOPCIDevice|VoodooI2CDeviceNub|VendorID|ProductID|HIDEventDriver  <class VoodooI2C|Interface  <class VoodooI2C' | cut -d "<" -f1
> ```
> 
> If there aren't any `ETPD` devices in the output, then the device itself hasn't been enabled in your DSDT. In either case though, I'd [dump your DSDT](https://dortania.github.io/Getting-Started-With-ACPI/Manual/dump.html#from-windows) on Windows so we can verify that it can correctly enable with the current SSDT-GPIO patch.
> 
> > > You would if it differs in the DSDT but it should match for all UX481FL variants since they use the same firmware.
> > 
> > 
> > I updated VoodooI2C.kext and used your VoodooI2CHID
> 
> You can also use the newer build of VoodooI2CHID from v2.9.1; my build back-ported changes to v2.8.

no results
![Uploading 20F627387558C9105BB05A4267AA04BD.jpg…]()

**Reply by Qonfused** [reply]
Do you get any output from the ioreg command? I can't see the image you uploaded there.

As mentioned before, I'd dump your DSDT and upload it here.

**Comment by MizhiDR** [comment]
![20F627387558C9105BB05A4267AA04BD](https://github.com/user-attachments/assets/11313892-b727-4a85-8f2b-f4c2ee499418)

**Reply by Qonfused** [reply]
It seems that the trackpad device simply isn't enabled in firmware (likely as it only gets enabled on Windows). Thankfully this is a relatively simple fix. All that is involved is flipping the device's `_STA` bit with an ACPI patch, afterwards it should work with either the original GPIO patches or the `-vi2c-force-polling` boot arg.

If you wouldn't mind [dumping your DSDT](https://dortania.github.io/Getting-Started-With-ACPI/Manual/dump.html#from-windows) from Windows and uploading it here, I can take a look at enabling your trackpad.

**Comment by MizhiDR** [comment]
[ioreg-ssdt.zip](https://github.com/user-attachments/files/18285836/ioreg-ssdt.zip)

This is my IOReg and dsdt file to trouble you

**Reply by Qonfused** [reply]
It looks like both the trackpad (`ETPD`) and the GPIO device are enabled, though there isn't a _CRS method for the `ETPD` device which is why VoodooI2C isn't attaching any drivers. VoodooI2C needs this method to determine what kind of input device it is, but it will otherwise skip attaching anything if there isn't one found.

![image](https://github.com/user-attachments/assets/3d3fd8fb-321d-4020-8d4e-6c4ec7fcc2ad)

For example, in the IOReg dump VoodooI2C attaches it's driver to the GPIO device (which appears just below it). When we enable the trackpad, we should see a **VoodooI2CPrecisionTouchpadHIDEventDriver** driver below the `ETPD` device in registry:

```bash
$ ioreg -rxn IOHIDInterface -k "VoodooI2CServices Supported" -t | grep -E 'IOACPIPlatformDevice|IOPCIDevice|VoodooI2CDeviceNub|VendorID|ProductID|HIDEventDriver  <class VoodooI2C|Interface  <class VoodooI2C' | cut -d "<" -f1
>>>      +-o PCI0@0  
>>>          +-o I2C1@15,1  
>>>                  +-o ETPD  
>>>                        |   "VendorID" = 0x4f3
>>>                        |   "ProductID" = 0x310e
>>>                        +-o VoodooI2CPrecisionTouchpadHIDEventDriver  
>>>                          +-o VoodooI2CMultitouchInterface  
```

Currently the original _CRS method is still renamed under the `ACPI > Patch` section of the config.plist file. You'll either want to comment them out of the `src/config.yml` file and re-build this project or disable them in the config.plist (`EFI/OC/config.plist`):
https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/blob/83efd22fafbbeaf76c5bec7a3fa868ed32c49908/src/config.yml#L131-L146
![image](https://github.com/user-attachments/assets/a5cb0cc0-6826-4cc2-b3bc-511abe07b1bd)

Alternatively, you can also create the _CRS method yourself and return the XCRS method's output if you plan to revisit GPIO pinning later:

(SSDT-GPIO.dsl)
```c++
/*
 * Enables GPIO device and pinning for Trackpad and Primary/Secondary displays.
 * @see https://github.com/VoodooI2C/VoodooI2C/blob/master/Documentation/GPIO%20Pinning.md
 *
 * Refer to the below issue comment for GPIO pin calculations:
 * https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/24#issuecomment-1399232818
 */
DefinitionBlock ("", "SSDT", 2, "UX481", "GPI0", 0x00000000)
{
    External (_SB_.PCI0.I2C1.ETPD, DeviceObj)
    External (GPHD, IntObj)
    External (USTP, IntObj)
    // config.plist ACPI > Patch renames
    External (_SB_.PCI0.I2C1.ETPD.XCRS, MethodObj)

    If (_OSI ("Darwin"))
    {
        // Enables _STA method in GPIO device
        GPHD = Zero
        // Enables correct SSCN and FMCN bus timings for each I2C controller.
        USTP = One
    }

    // GPIO - Trackpad
    Scope (_SB.PCI0.I2C1.ETPD)
    {        
        Return (XCRS ())
    }
}
```

**Reply by MizhiDR** [reply]
> 看起来触控板 ( `ETPD`) 和 GPIO 设备都已启用，但该设备没有 _CRS 方法，`ETPD`这就是 VoodooI2C 没有连接任何驱动程序的原因。VoodooI2C 需要此方法来确定它是什么类型的输入设备，但如果没有找到输入设备，它将跳过连接任何设备。
> 
> ![图像](https://private-user-images.githubusercontent.com/32466081/399574105-3d3fd8fb-321d-4020-8d4e-6c4ec7fcc2ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4MDE2NTgsIm5iZiI6MTczOTgwMTM1OCwicGF0aCI6Ii8zMjQ2NjA4MS8zOTk1NzQxMDUtM2QzZmQ4ZmItMzIxZC00MDIwLThkNGUtNmM0ZWM3ZmNjMmFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE3VDE0MDkxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlYzM0ZTgyNWM3MmMyNDkzNzc1NWE4YTNhZTllMTRiYjk0MTg4N2M1ZjM2NmU5YTJiZGFiMDIyMjRjYmQwYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.y1PXDONGoczGF21RShtptLsFY61avTsrc5neOvYEqPE)
> 
> 例如，在 IOReg 转储中，VoodooI2C 将其驱动程序附加到 GPIO 设备（该设备就出现在其下方）。当我们启用触控板时，我们应该在注册表中的设备下方看到**VoodooI2CPrecisionTouchpadHIDEventDriver**`ETPD`驱动程序：
> 
> ```shell
> $ ioreg -rxn IOHIDInterface -k "VoodooI2CServices Supported" -t | grep -E 'IOACPIPlatformDevice|IOPCIDevice|VoodooI2CDeviceNub|VendorID|ProductID|HIDEventDriver  <class VoodooI2C|Interface  <class VoodooI2C' | cut -d "<" -f1
> >>>      +-o PCI0@0  
> >>>          +-o I2C1@15,1  
> >>>                  +-o ETPD  
> >>>                        |   "VendorID" = 0x4f3
> >>>                        |   "ProductID" = 0x310e
> >>>                        +-o VoodooI2CPrecisionTouchpadHIDEventDriver  
> >>>                          +-o VoodooI2CMultitouchInterface  
> ```
> 
> 目前，原始的 _CRS 方法仍在`ACPI > Patch`config.plist 文件的部分下重命名。您可以将其从`src/config.yml`文件中注释掉并重新构建此项目，或者在 config.plist 中禁用它们（`EFI/OC/config.plist`）：
> 
> https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/blob/83efd22fafbbeaf76c5bec7a3fa868ed32c49908/src/config.yml#L131-L146
> 
> 
> ![图像](https://private-user-images.githubusercontent.com/32466081/399574469-a5cb0cc0-6826-4cc2-b3bc-511abe07b1bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4MDE2NTgsIm5iZiI6MTczOTgwMTM1OCwicGF0aCI6Ii8zMjQ2NjA4MS8zOTk1NzQ0NjktYTVjYjBjYzAtNjgyNi00Y2MyLWIzYmMtNTExYWJlMDdiMWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE3VDE0MDkxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1ODM3OGQxOGE3MDUyMmEzY2UwMTFhMzlhYmFjNzhiZTNmYWFjZmNkYTZhNTk5MTFhNDNjNDE2Y2RmM2FhMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._lhyzkgy_Dr_pTpYTBr8-la9h--cT6M6feP7CCb3Ezw)
> 或者，如果您计划稍后重新访问 GPIO 固定，您也可以自己创建 _CRS 方法并返回 XCRS 方法的输出：
> 
> （SSDT-GPIO.dsl）
> 
> ```c++
> /*
>  * Enables GPIO device and pinning for Trackpad and Primary/Secondary displays.
>  * @see https://github.com/VoodooI2C/VoodooI2C/blob/master/Documentation/GPIO%20Pinning.md
>  *
>  * Refer to the below issue comment for GPIO pin calculations:
>  * https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/issues/24#issuecomment-1399232818
>  */
> DefinitionBlock ("", "SSDT", 2, "UX481", "GPI0", 0x00000000)
> {
>     External (_SB_.PCI0.I2C1.ETPD, DeviceObj)
>     External (GPHD, IntObj)
>     External (USTP, IntObj)
>     // config.plist ACPI > Patch renames
>     External (_SB_.PCI0.I2C1.ETPD.XCRS, MethodObj)
> 
>     If (_OSI ("Darwin"))
>     {
>         // Enables _STA method in GPIO device
>         GPHD = Zero
>         // Enables correct SSCN and FMCN bus timings for each I2C controller.
>         USTP = One
>     }
> 
>     // GPIO - Trackpad
>     Scope (_SB.PCI0.I2C1.ETPD)
>     {        
>         Return (XCRS ())
>     }
> }
> ```

Thank you for your help last time. I have a question for you: how do you drive the screen touch?

**Reply by Qonfused** [reply]
They use the VoodooI2CHID + VoodooI2C drivers, which provides it's own touchscreen driver. It is a rather unusual setup for VoodooI2C's logic, so I've created an auxiliary kext that ensures the correct drivers attach to each display and the trackpad:
https://github.com/Qonfused/VoodooInjector

I have been unable to find a stable GPIO setup for the two displays though (and they will eventually time out when in normal I2C mode), so at the moment they're configured to use APIC pins which are only somewhat stable:
https://github.com/Qonfused/ASUS-ZenBook-Duo-14-UX481-Hackintosh/blob/83efd22fafbbeaf76c5bec7a3fa868ed32c49908/src/ACPI/SSDT-GPIO.dsl#L58-L144

**Reply by MizhiDR** [reply]
Thank you for your ideas. Do you have any contact information? I would like to ask you some questions individually.
